### PR TITLE
feat: read null sort_key_ids

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes https://github.com/influxdata/influxdb_iox/issues/8480

Due to the incident last Friday https://github.com/influxdata/idpe/issues/18019 that lead to the revert https://github.com/influxdata/influxdb_iox/pull/8477, I want to be cautious this time. This PR is read NULL `sort_key_ids` and add a lot of tests to verify the NULL values.

1. The tests can be more than needed but they do not only help us to avoid ColumnDecode issue but will also help in next few PRs to verify empty  array,`{}`,  for insert tests and has values for update tests.
2. I want to ask the reviewers about ColumnDecode of NULL array. I think it should work since all tests work but I want to hear from others.

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: ingester/src/buffer_tree/partition.rs [hunk 1]
```diff
@@ -1001,14 +1001,18 @@ mod tests {
             .create_or_get("test".into(), table_id)
             .await
             .expect("should create");
+        // sort key is null
+        assert!(partition.sort_key_ids.is_none());
```
Comment (NGA-TRAN) [inline ingester/src/buffer_tree/partition.rs:1005]: For this PR, all test partitions will be verified with NULL (None) `sort_key_ids`. 
  . Next PR that modifies insert partitions, they will be empty array `{}` for both insert and update partition test cases.
  . A PR later that updates partition's `sort_key_ids`, the value of test sort_key_ids will be `{}` for insert partitions and have specific values for update partitions.

So all of verification NULL asserts in this PR will be valuable for verifying different values in the next few PRs. This is the key for NOT hitting sqlx's ColumnDecode issue that we hit last Friday

File: data_types/src/partition.rs [hunk 2]
```diff
@@ -383,7 +391,7 @@ pub struct Partition {
     /// For example, updating `A,B,C` to either `A,D,B,C` or `A,B,C,D`
     /// is legal. However, updating to `A,C,D,B` is not because the
     /// relative order of B and C have been reversed.
-    pub sort_key: Vec<String>,
+    pub sort_key_ids: Option<ColumnSet>,
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply data_types/src/partition.rs:394]: So this is one way to handle the problem. Another way is to set this to `ColumnSet` but annotate the field with [`#[sqlx(default)]`](https://docs.rs/sqlx/latest/sqlx/trait.FromRow.html#default), and if `sort_key_ids` isn't present in the query results, it'll get the default `ColumnSet` (which is `ColumnSet(vec![])`).

Another solution could be to write a migration that does:

```
UPDATE partition
SET sort_key_ids = '{}';
```

and then in another deploy, after verifying that there are no partitions that have `NULL` `sort_key_ids`, run another migration that requires `sort_key_ids` not be NULL. 

Is there a reason for choosing this solution over the others?

File: iox_catalog/src/interface.rs [hunk 3]
```diff
@@ -1647,6 +1666,8 @@ pub(crate) mod test_helpers {
             updated_other_partition.sort_key,
             vec!["tag2", "tag1", "time"]
         );
+        // sort_key_ids should be null
+        assert!(updated_other_partition.sort_key_ids.is_none());
```
Comment (carols10cents) [inline iox_catalog/src/interface.rs:1670]: I don't think adding these assertions everywhere is really providing much value or assurance that this isn't going to crash; we could assert lots of things that we don't expect to change all over the place, but that isn't really providing value. This seems like it's adding lots of places that will need to be updated, maintained, or deleted in the future....

File: iox_catalog/src/postgres.rs [hunk 4]
```diff
@@ -1461,10 +1486,12 @@ RETURNING *
         .context(interface::CouldNotDeleteSkippedCompactionsSnafu)
     }
 
+    // todo: test return null sort_key_ids
     async fn most_recent_n(&mut self, n: usize) -> Result<Vec<Partition>> {
         sqlx::query_as(
+// TODO: ask Carol where persisted_sequence_number is stored because there is no such filed in Partition struct
```
Comment (carols10cents) [inline iox_catalog/src/postgres.rs:1492]: It's not stored anywhere; it just gets ignored. `persisted_sequence_number` can probably be removed from this `SELECT` now, but in a separate PR.

File: data_types/src/partition.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (carols10cents) [reply data_types/src/partition.rs:394]: Related: https://github.com/influxdata/influxdb_iox/pull/4801

[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply data_types/src/partition.rs:394]: So her are the option:
1. This PR solution `pub sort_key_ids: Option<ColumnSet>`
2. Carol's solution 1:  `pub sort_key_ids: ColumnSet>` plus [#[sqlx(default)]](https://docs.rs/sqlx/latest/sqlx/trait.FromRow.html#default), and if sort_key_ids isn't present in the query results, it'll get the default ColumnSet (which is ColumnSet(vec![])).
3.  `ALTER partition ALTER COLUM  sort_key_ids SET NOT NULL  default {}`. This won't affect catalog performance becasue the default won't get updated to the existing data at all. Postgres only do it when the data rows are used

I do not think the `UPDATE partition ...` is a choice because there are always new partitions created and sort_key_ids  end up to be NULL until a few more tickets we plan to do is done.

I vote for either first (this PR) or the third option.

File: iox_catalog/src/postgres.rs [hunk 5]
```diff
@@ -1157,6 +1157,31 @@ impl PartitionRepo for PostgresTxn {
...
+        //     to simulate existing data with NULL sort_key_ids for me to tests sort_key_ids with None or Some(ColumnSet)
+        //   . In another PR, I will update sort_key_ids of partitions with coming data with corresponding sort_key, and verify
+        //     those sort_key_ids are Some(ColumnSet) with non-empty ColumnSet. At this time, our code will handle 3 kinds of
+        //    sort_key_ids: None, Some(ColumnSet) with empty ColumnSet, Some(ColumnSet) with non-empty ColumnSet.
+        //   . Then in near future after Marco finishes his script to update all NULL sort_key_ids with corresponding sort_key,
+        //     we will make sort_key_ids NOT NULL and then we will  modify the code to not accept None sort_key_ids anymore.
+        //
+        // HOwever, if sort_key_ids must be a empty array, {}, rather than null, we can change the catalog field to NOT NULL and make it default `{}`.
+        // Unless the NULL does not work with sqlx's `ColumnDecode`, I think we should keep it as NULL for now.
+        // At later step, we will convert it to NOT NULL when we know that all existing data has been updated.
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply iox_catalog/src/postgres.rs:1184]: If we decide to go this direction rather than doing a migration that inserts empty array and then sets a NOT NULL constraint, take a look at some tests I've added for cases like this-- they have to be added to postgres/sqlite specifically, because the application won't let you create records in the old state so you have to put raw SQL in the tests. Here's one that makes a partition without a hash id: https://github.com/influxdata/influxdb_iox/blob/964b2f6b9737df75cd0f4e1cee96e03ce70a4d40/iox_catalog/src/postgres.rs#L2192-L2210

File: iox_catalog/src/interface.rs (hunk 3 above)
Comment (NGA-TRAN) [reply iox_catalog/src/interface.rs:1670]: While I agree there are a lot of asserts and we may not need all, I do think we need at least 70% of these asserts for the next few PRs where the NULL will be either empty array `{}` or arrays with actual values. Thus they look redundant in this PR, they will be very valuable in the next few PRs. I would leave all of them here and I will remove the redundant ones in the next few PRs after they are converted to `{}` or arrays with actual values

File: data_types/src/partition.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply data_types/src/partition.rs:394]: Thinking more. NULLvalue is the trouble maker that has caused all trouble. **We should avoid that in the first place** if possible and this case we can because we have not used `sort_key_ids` yet. I propose we go with **option 3** above. 

I will

1.  Have a PR to do `ALTER partition ALTER COLUM  sort_key_ids SET NOT NULL  default {}`.
2. Then I will modify here to make it `pub sort_key_ids: ColumnSet,`
3. Then I can remove a lot of tests here. Only keep some representative tests.

[the remark this answers is not in the document]
Comment (crepererum) [reply data_types/src/partition.rs:394]: Please do NOT use `default`. There's a reason that NULL is THE postgres marker for missing data and using a sentinel value makes debugging, migrating and reading the data way harder. Please handle NULLs in the Rust code base.

File: data_types/src/columns.rs [hunk 6]
```diff
@@ -379,6 +379,15 @@ impl Deref for ColumnSet {
     }
 }
 
+impl<I> From<I> for ColumnSet
+where
+    I: IntoIterator<Item = i64>,
+{
+    fn from(ids: I) -> Self {
+        Self(ids.into_iter().map(ColumnId::new).collect())
```
Comment (crepererum) [inline data_types/src/columns.rs:387]: This is invalid because it breaks Eq and Co. Internally the data is sorted, so please just use the `new` method that does this:

```suggestion
        Self::new(ids.into_iter().map(ColumnId::new).collect())
```

File: data_types/src/partition.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply data_types/src/partition.rs:394]: Let us stick with option 1 that handle NULL (this PR) then

File: iox_catalog/src/postgres.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply iox_catalog/src/postgres.rs:1184]: Let me explore @carols10cents's comment above to understand and use it

[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply iox_catalog/src/postgres.rs:1184]: @carols10cents 2 tests with same name `existing_partitions_without_sort_key_ids` have been added for Postgres and sqlite

**Review by NGA-TRAN — COMMENTED** [review]
@carols10cents and @crepererum : I have addressed all your comments. Can you re-review this? If this is approved by by 2pm ET today, I will merge it and observe how it works on Serverless. Otherwise, I will merge it first thing tomorrow

File: data_types/src/columns.rs [hunk 7]
```diff
@@ -331,7 +331,7 @@ impl TryFrom<proto::column_schema::ColumnType> for ColumnType {
 }
 
 /// Set of columns.
-#[derive(Debug, Clone, PartialEq, Eq, sqlx::Type)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Default, sqlx::Type)]
```
Comment (crepererum) [inline data_types/src/columns.rs:334]: Is `Defaults` actually needed here?

File: data_types/src/partition.rs (hunk 2 above)
Comment (crepererum) [inline data_types/src/partition.rs:394]: `ColumnSet` is the WRONG type here. It's a set (and internally normalizes the order), you want a "sorted set". Please introduce as new type `SortedColumnSet`.

Sorry, I should have caught this during the first round.

**Review by crepererum — CHANGES_REQUESTED** [review]

File: iox_catalog/src/interface.rs (hunk 3 above)
Comment (carols10cents) [reply iox_catalog/src/interface.rs:1670]: So the part I'm not understanding is, under what circumstances would, say, only 1 or 2 out of these assertions would fail but not all of them? Put another way, what problem(s) are these assertions guarding against?

If they're guarding against having an incorrect initial database state, then we only need one of these, because if one fails, then they'll all fail. 

Is it likely that a code change in the next few PRs will spuriously make, say, `PartitionRepo::get_by_id` but NOT `PartitionRepo::create_or_get` return partitions with non-`None` `sort_key_ids`? That doesn't seem very likely to me. I'm just having trouble imagining the failure scenarios here.

File: data_types/src/partition.rs (hunk 2 above)
Comment (carols10cents) [reply data_types/src/partition.rs:394]: Why does `ColumnSet` hold a `Vec` rather than a `HashSet` actually? I was going to submit a PR to change its name when I saw it was a `Vec` because I didn't look closely at `ColumnSet::new` where the actual set implementation is.

File: data_types/src/columns.rs (hunk 7 above)
Comment (NGA-TRAN) [reply data_types/src/columns.rs:334]: You eyes catch things better than auto clips 🙂 . Everything works after removing `Default`

File: data_types/src/partition.rs (hunk 2 above)
Comment (NGA-TRAN) [reply data_types/src/partition.rs:394]: @crepererum `ColumnSet` is define as: `pub struct ColumnSet(Vec<ColumnId>);` which I think `ColumnIds` will stay in the order of `Vec<ColumnId>` and are sorted. In the PR that was reverted last Friday and I am working on it locally right now, I had tests that show their orders.

If you do not think the `pub struct ColumnSet(Vec<ColumnId>);` is the right way, what is the `SortColumnSet` looks like?

Actually, @carols10cents is the one who introduced `ColumnSet` in this case and I have agreed. What do you think Carol?

Comment (NGA-TRAN) [reply data_types/src/partition.rs:394]: Haha, Carol and I added comment concurrently

File: iox_catalog/src/interface.rs (hunk 3 above)
Comment (NGA-TRAN) [reply iox_catalog/src/interface.rs:1670]: @carols10cents See [my second PR here](https://github.com/influxdata/influxdb_iox/pull/8494) that depends on this PR addressed your concern. I only keep necessary tests and also added comments for those tests. Actually this add-and-remove-with-comments exercise helps me a lot to understand the needed tests and available tests.

File: iox_catalog/src/postgres.rs [hunk 8]
```diff
@@ -2239,6 +2270,77 @@ RETURNING id, hash_id, table_id, partition_key, sort_key, new_file_at;
...
+        .bind(table_id) // $3
+        .fetch_one(&pool)
+        .await
+        .unwrap();
+
+        // Check that the sort_key_ids is null in the database doesn't break querying for partitions.
+        let table_partitions = repos.partitions().list_by_table_id(table_id).await.unwrap();
+        assert_eq!(table_partitions.len(), 1);
+        let partition = &table_partitions[0];
+        assert!(partition.hash_id().is_none());
```
Comment (carols10cents) [inline iox_catalog/src/postgres.rs:2313]: The assertions on the hash_id aren't relevant here; those were only relevant for the `hash_id` migration tests.

File: iox_catalog/src/postgres.rs [hunk 9]
```diff
@@ -2214,6 +2243,8 @@ RETURNING id, hash_id, table_id, partition_key, sort_key, new_file_at;
         assert_eq!(table_partitions.len(), 1);
         let partition = &table_partitions[0];
         assert!(partition.hash_id().is_none());
+        // assert null sort_key_ids
+        assert!(partition.sort_key_ids.is_none());
```
Comment (carols10cents) [inline iox_catalog/src/postgres.rs:2247]: And this`sort_key_ids` assertion isn't relevant for the `hash_id` test. These are more unit-style tests and should only be testing one thing. In fact, this test should probably be inserting an empty array for `sort_key_ids` to accurately test a partition that doesn't have a hash ID in the catalog but DOES have `sort_key_ids` set.

File: data_types/src/partition.rs (hunk 2 above)
Comment (crepererum) [reply data_types/src/partition.rs:394]: `ColumnSet` is a SET (as the type says) and the fact that it is internally using a `Vec` is irrelevant for its semantic. This type is already used for the parquet file entries and there the ordering doesn't matter (however Postgres doesn't have a set type, so the typical way to implement this is by using a sort-normalized vector/array).

File: iox_catalog/src/postgres.rs [hunk 10]
```diff
@@ -2173,6 +2200,8 @@ mod tests {
             .unwrap();
         assert_eq!(table_partitions.len(), 1);
         assert_eq!(table_partitions[0].hash_id().unwrap(), &hash_id);
+        // sort_key_ids is null
+        assert!(table_partitions[0].sort_key_ids.is_none());
```
Comment (carols10cents) [inline iox_catalog/src/postgres.rs:2204]: To me, this assertion here covers all the other assertions added after a call to `PartitionRepo::create_or_get` because if something is wrong in `create_or_get`, this test will fail. All the others seem redundant to this one, to me.

File: data_types/src/partition.rs (hunk 2 above)
Comment (NGA-TRAN) [reply data_types/src/partition.rs:394]: @crepererum, @carols10cents and I have a quick chat and we agree on creating a new type `SortedColumnSet` for this that:
1. Type: pub struct SortedColumnSet(Vec<ColumnId>);
2. Its `new` function will keep the column order of the input

I am working on this

File: iox_catalog/src/postgres.rs (hunk 10 above)
Comment (NGA-TRAN) [reply iox_catalog/src/postgres.rs:2177]: Yeah, there should a few places for different purpose that I added in the comments

File: data_types/src/partition.rs (hunk 2 above)
Comment (NGA-TRAN) [reply data_types/src/partition.rs]: Done

File: iox_catalog/src/postgres.rs [hunk 11]
```diff
@@ -1157,6 +1157,31 @@ impl PartitionRepo for PostgresTxn {
         // Note: since sort_key is now an array, we must explicitly insert '{}' which is an empty
         // array rather than NULL which sqlx will throw `UnexpectedNullError` while is is doing
         // `ColumnDecode`
+        //
+        // todo: ask Carol/Dom/Andrew/Marco whether the note above (empty array `{}`) is a must becasue this PR does not follow this for `sort_key_ids`.
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply iox_catalog/src/postgres.rs]: I also remove the no-longer-valid existing comments

File: iox_catalog/src/postgres.rs (hunk 9 above)
Comment (NGA-TRAN) [reply iox_catalog/src/postgres.rs]: Since this PR does not yet insert anything into sort_key_ids, I will add empty sort_key_ids in this test in my next PR that implements empty sort_key_ids

File: data_types/src/columns.rs [hunk 12]
```diff
@@ -379,6 +380,63 @@ impl Deref for ColumnSet {
...
+    pub fn new<I>(columns: I) -> Self
+    where
+        I: IntoIterator<Item = ColumnId>,
+    {
+        let mut columns: Vec<ColumnId> = columns.into_iter().collect();
+
+        let len_pre_dedup = columns.len();
+        columns.dedup();
+        let len_post_dedup = columns.len();
+        assert_eq!(len_pre_dedup, len_post_dedup, "set contains duplicates");
```
Comment (crepererum) [inline data_types/src/columns.rs:404]: IIRC `Vec::dedup` only works on sort-normalized vectors. However you don't wanna mess with the sort order. I see two options: remove the check or make it slightly more advanced:

```rust
let mut columns_sorted = columns.clone();
columns_sorted.sort();
let len_pre_dedup = columns_sorted.len();
columns_sorted.dedup();
let len_post_dedup = columns_sorted.len();
assert_eq!(len_pre_dedup, len_post_dedup, "set contains duplicates");
```

and then continue with the normal `columns` (NOT w/ `columns_sorted`!):

```rust
columns.shrink_to_fit();
Self(columns)
```

**Review by crepererum — APPROVED** [review]
Small nitpick regarding the duplication check, this probably doesn't need another round of reviews after fixing.

File: data_types/src/columns.rs (hunk 12 above)
Comment (NGA-TRAN) [reply data_types/src/columns.rs]: Used the advanced one and al added 2 tests for it according
